keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -193,123 +336,634 @@ each list item is made of:
+
+This means two things:
+1. when the server receives a request with a discriminator set to a
+given field, and multiple member fields are set it should:
+  a. If the discriminator has been modified from the one set in the old object,
+  the server should clear all fields except the one pointed to by the discriminator.
+  b. If the discriminator is unchanged but a new member field has been set (in
+  addition to the existing one that has not been cleared), it should error
+  loudly that the client must change the discriminator if it changes any union
+  member fields.
+2. when the server receives a request with a discriminator set to a given field,
+   but that given field is empty, the server should fail with a clear error
+   message.
+
+For situations where a typed client is unaware of a new field (that is currently
+set) and drops the set field such that no fields are now set, the server will
+clear all the fields of the union. This is unavoidable for unions without a
+discriminator and a major reason why we recommend all new unions have a
+discriminator.
+
+Union types without a discriminator must continue to perform normalization and
+validation manually (i.e. per resource in the validation functions), as is
+currently done for existing union types.
+
+Objects must be validated AFTER the normalization process.
+
+### Open Questions
+
+A few open questions exist with the design that need to be resolved with SIG API
+Machinery before moving forward.
+
+1. What do we do when the discriminator is set (to the previously set value),
+   but all union member fields are null? This likely implies that the client cleared
+   the union member unintentionally, given that the discriminator value remains.
+   Should the server:
+   a. Error loudly and inform the client that the field pointed to by the
+   discriminator is null?
+   b. Retain the previously set member field, present on the old object, absent
+   from the new object, but pointed to by the discriminator?
+2. What should the value of the discriminator be when no field in the union is
+   to be set (i.e for optional unions).
+   a. mandate that the discriminator is the empty string for all optional unions?
+   b. make the field specified on a per union basis (defined in the go markers)?
+   c. mandate that the discriminator be unset (thus make the discriminator an
+   optional field for optional unios)?
+3. How should a server behave when it receives a union with multiple fields set
+   and the discriminator modified to point to the newly set member field?
+   a. reject the request,
+   b. accept the request but don't clear any fields
+   c. accept the request and clear all the fields except the one chosen by the
+   discriminator
+4. Given that we want to migrate at least one existing union to use the new
+   semantics, which union should we do? Must we upgrade both a discriminated and
+   non-discriminated union for alpha or is only upgrading an existing
+   discriminated union sufficient?
```

Comment:
one of each, to discover flaws in the design.

Reply:
That makes sense, but I think there's argument to be made that we shouldn't even try migrating unions that don't use a discriminator and just accept that they will always use the existing manual validations? thoughts.

I've also added a discussion on some proposed initial types to migrate. It's a little shallow, so I'm open to any improvements on a good methodology for choosing the first types to migrate.